onnxruntime/core/providers/nnapi/nnapi_builtin/nnapi_api_helper.cc

```diff
@@ -0,0 +1,120 @@
+#include "nnapi_api_helper.h"
+
+#include "core/providers/nnapi/nnapi_builtin/builders/model_builder.h"
+#include "core/providers/nnapi/nnapi_builtin/nnapi_lib/nnapi_implementation.h"
+#include "core/common/logging/logging.h"
+
+#ifdef __ANDROID__
+#include <android/api-level.h>
+#endif
+
+namespace onnxruntime {
+namespace nnapi {
+
+int32_t GetNNAPIRuntimeFeatureLevel(const ::NnApi* nnapi) {
+  if (!nnapi)
+    return 0;
+  int32_t runtime_level = static_cast<int32_t>(nnapi->nnapi_runtime_feature_level);
+#ifdef __ANDROID__
+  int device_api_level = android_get_device_api_level();
+  runtime_level = (device_api_level < __ANDROID_API_S__) ? device_api_level : runtime_level;
+#endif
+  return runtime_level;
+}
+
+int32_t GetDeviceFeatureLevel(const ::NnApi* nnapi, const std::vector<ANeuralNetworksDevice*>& device_handles) {
+  if (!nnapi)
+    return 0;
+
+  int32_t target_feature_level = GetNNAPIRuntimeFeatureLevel(nnapi);
+
+  int64_t devices_feature_level = -1;
+
+  for (const auto* device_handle : device_handles) {
+    int64_t curr_device_feature_level = 0;
+    if (nnapi->ANeuralNetworksDevice_getFeatureLevel(device_handle, &curr_device_feature_level) != ANEURALNETWORKS_NO_ERROR) {
+      continue;
+    }
+
+    devices_feature_level = std::max(curr_device_feature_level, devices_feature_level);
+  }
+
+  if ((devices_feature_level > 0) && (devices_feature_level < target_feature_level)) {
+    LOGS_DEFAULT(INFO) << "Changing NNAPI Feature Level " << target_feature_level
+                       << " to supported by target devices: " << devices_feature_level;
+
+    target_feature_level = static_cast<int32_t>(devices_feature_level);
+  }
+  return target_feature_level;
+}
+
+[[nodiscard]] Status GetTargetDevices(const ::NnApi* nnapi, TargetDeviceOption target_device_option,
```

Comment:
nit: Status is already marked nodiscard so we don't need the attribute on functions returning Status

Reply:
Got it. Thanks